airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_requester.py

```diff
@@ -155,3 +168,343 @@ def cache_filename(self) -> str:
+        if hasattr(self.error_handler, "max_retries"):
+            return self.error_handler.max_retries
+        return self._DEFAULT_MAX_RETRY
+
+    @property
+    def logger(self):
+        return logging.getLogger(f"airbyte.HttpRequester.{self.name}")
+
+    def _should_retry(self, response: requests.Response) -> bool:
+        """
+        Specifies conditions for backoff based on the response from the server.
+
+        By default, back off on the following HTTP response statuses:
+         - 429 (Too Many Requests) indicating rate limiting
+         - 500s to handle transient server errors
+
+        Unexpected but transient exceptions (connection timeout, DNS resolution failed, etc..) are retried by default.
+        """
+        if self.error_handler is None:
+            return response.status_code == 429 or 500 <= response.status_code < 600
+        return self.interpret_response_status(response).action == ResponseAction.RETRY
+
+    def _backoff_time(self, response: requests.Response) -> Optional[float]:
+        """
+        Specifies backoff time.
+
+         This method is called only if should_backoff() returns True for the input request.
+
+         :param response:
+         :return how long to backoff in seconds. The return value may be a floating point number for subsecond precision. Returning None defers backoff
+         to the default backoff behavior (e.g using an exponential algorithm).
+        """
+        if self.error_handler is None:
+            return None
+        should_retry = self.interpret_response_status(response)
+        if should_retry.action != ResponseAction.RETRY:
+            raise ValueError(f"backoff_time can only be applied on retriable response action. Got {should_retry.action}")
+        assert should_retry.action == ResponseAction.RETRY
+        return should_retry.retry_in
+
+    def _error_message(self, response: requests.Response) -> str:
+        """
+        Constructs an error message which can incorporate the HTTP response received from the partner API.
+
+        :param response: The incoming HTTP response from the partner API
+        :return The error message string to be emitted
+        """
+        return self.interpret_response_status(response).error_message
+
+    def _get_mapping(self, method, **kwargs):
+        """
+        Get mapping from the provided method, and get the keys of the mapping.
+        If the method returns a string, it will return the string and an empty set.
+        If the method returns a dict, it will return the dict and its keys.
+        """
+        mapping = method(**kwargs) or {}
+        keys = set(mapping.keys()) if not isinstance(mapping, str) else set()
+        return mapping, keys
+
+    def _get_request_options(
+        self,
+        stream_slice: Optional[StreamSlice],
+        next_page_token: Optional[Mapping[str, Any]],
+        requester_method,
+        auth_options_method,
+        extra_options: Optional[Mapping[str, Any]] = None,
```

Comment:
it looks like extra_options can also be a string if this method is called from [_request_body_data](https://github.com/airbytehq/airbyte/pull/28050/files#diff-17485be091db6243dff1b4ef8d77a71b199fd37a09be7cbaa3ebaaea4228ed29R315). Is that expected?

Reply:
Yes, I just forgot to adjust the type. If it's a string it will make sure there is only one and throw otherwise.